ga4gh/server/backend.py

```diff
@@ -138,6 +172,268 @@ def datasetsGenerator(self, request, access_map):
+    def logicHandler(self, logic, responses, dataset_id):
+        """
+        :param  logic: dict parsed from query containing logic statement keys or component id keys
+                responses: object with key being the id, and the value being the response from corresponding endpoints
+        :return: list of patient_id filtered based on join logic
+        """
+
+        op_keys = ['and', 'or']
+
+        if len(logic.keys()) == 1:
+            logic_key = logic.keys()[0]
+        else:
+            # too many logic keys
+            raise exceptions.InvalidLogicException('Invalid number of keys')
+
+        if logic_key in op_keys:
+            results_arr = []
+            patient_array = []
+
+            for logic_obj in logic[logic_key]:
+                results_arr.append(self.logicHandler(logic_obj, responses, dataset_id))
+
+            if logic_key == 'or':
+                for id_list in results_arr:
+                    for id in id_list:
+                        if id not in patient_array:
+                            patient_array.append(id)
+            elif logic_key == 'and':
+                results_arr.sort(key=len)
+                for id in results_arr[0]:
+                    if all(id in results_arr[x] for x in range(1, len(results_arr))):
+                        patient_array.append(id)
+
+            return patient_array
+
+        elif logic_key == 'id':
+            id_list = []
+            for response in responses[logic[logic_key]]:
+                patient_id = self.getResponsePatientId(response, dataset_id)
+                if patient_id not in id_list:
+                    id_list.append(patient_id)
+
+            return id_list
+
+        else:
+            # invalid logic key
+            raise exceptions.InvalidLogicException("Invalid key used")
+
+    def getResponsePatientId(self, response, dataset_id):
+        """
+        Gets the patientId from the response for object joins, otherwise throw error
+        :param response:
+        :param dataset_id:
+        :return: patient id string
+        """
+        if 'patientId' in response:
+            return response['patientId']
+        elif 'variantSetId' in response:
+            dataset = self.getDataRepository().getDataset(dataset_id)
+            variantSet = dataset.getVariantSet(response['variantSetId'])
+            return variantSet.getPatientId()
+        else:
+            raise exceptions.BadRequestException
+
+    def componentsHandler(self, datasetId, components, return_mimetype, access_map):
+        """
+        Parse the component portion of incoming request
+        :param datasetId;
+        :param components:
+        :param access_map
+        :return: responses object with key being the id, and the value being the response from corresponding endpoints
+        """
+        requests = {}
+        idMapper = {}
+
+        for component in components:
+            keyList = list(component.keys())
+            if len(keyList) != 2:
```

Comment:
Does it enforce that the number of components has to be two? We should allow to have components two or more.

Reply:
I think it is just a soft check that each component can only have 2 keys : id and filter